### ISSUE TITLE
[JENKINS-60625] Dashboard rendering blocked on UpdateSite.getJSONObject

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#9437]
<p>Recently I have noticed Jenkins page rendering being blocked for a long time (sometimes nearly 30s) with stack traces such as</p>

<pre>
"Handling GET /jenkins/ from ... : ... View/index.jelly AdministrativeMonitorsDecorator/footer.jelly" ...
   java.lang.Thread.State: RUNNABLE
	at java.lang.String.<init>(String.java:207)
	at java.lang.String.substring(String.java:1933)
	at net.sf.json.util.JSONTokener.matches(JSONTokener.java:110)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:912)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONArray._fromJSONTokener(JSONArray.java:1131)
	at net.sf.json.JSONArray.fromObject(JSONArray.java:125)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:351)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject._fromString(JSONObject.java:1145)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:162)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:132)
	at hudson.model.UpdateSite.getJSONObject(UpdateSite.java:338)
	at hudson.model.UpdateSite.getData(UpdateSite.java:323)
	at hudson.model.UpdateCenter.getCoreSource(UpdateCenter.java:604)
	at hudson.model.UpdateCenter$CoreUpdateMonitor.getData(UpdateCenter.java:1035)
	at hudson.model.UpdateCenter$CoreUpdateMonitor.isActivated(UpdateCenter.java:1030)
	at jenkins.model.Jenkins.lambda$getActiveAdministrativeMonitors$0(Jenkins.java:2183)
	at jenkins.model.Jenkins$$Lambda$103/1298422401.test(Unknown Source)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
	at jenkins.model.Jenkins.getActiveAdministrativeMonitors(Jenkins.java:2188)
	at jenkins.management.AdministrativeMonitorsDecorator.getActiveAdministrativeMonitors(AdministrativeMonitorsDecorator.java:77)
	at jenkins.management.AdministrativeMonitorsDecorator.getActiveAdministrativeMonitorsCount(AdministrativeMonitorsDecorator.java:72)
	at jenkins.management.AdministrativeMonitorsDecorator.shouldDisplay(AdministrativeMonitorsDecorator.java:131)
</pre>

<p>or</p>

<pre>
  java.lang.Thread.State: RUNNABLE
	at java.util.regex.Pattern.compile(java.base@​11.0.5/Pattern.java:1757)
	at java.util.regex.Pattern.<init>(java.base@​11.0.5/Pattern.java:1428)
	at java.util.regex.Pattern.compile(java.base@​11.0.5/Pattern.java:1068)
	at net.sf.json.regexp.JdkRegexpMatcher.<init>(JdkRegexpMatcher.java:38)
	at net.sf.json.regexp.JdkRegexpMatcher.<init>(JdkRegexpMatcher.java:31)
	at net.sf.json.regexp.RegexpUtils.getMatcher(RegexpUtils.java:39)
	at net.sf.json.util.JSONTokener.matches(JSONTokener.java:111)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:912)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONArray._fromJSONTokener(JSONArray.java:1131)
	at net.sf.json.JSONArray.fromObject(JSONArray.java:125)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:351)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject._fromString(JSONObject.java:1145)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:162)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:132)
</pre>

<p>While loading the JSON text from disk takes just milliseconds, the parsing is extremely slow (hundreds of times slower than <tt>jq</tt>, for example). This seems to be due to poor code in <tt>json-lib</tt>, especially <a href="https://github.com/jenkinsci/json-lib/blame/master/src/main/jdk15/net/sf/json/JSONObject.java#L829" class="external-link" target="_blank" rel="nofollow noopener">this</a> which repeatedly compiles patterns where a simple <tt>.startsWith("null")</tt> would apparently have sufficed!</p>

<p>The poor performance of <tt>JSONObject.fromObject</tt> might not have been so noticeable except that some administrative monitors blocked on the parsing, contrary to the Javadoc in <tt>AdministrativeMonitor.isActivated</tt>.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-60625" target="_blank">Dashboard rendering blocked on UpdateSite.getJSONObject</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>label(s)</b>: 2.204.2-rejected, 2.204.3-fixed, performance
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2020-01-04T14:22:14+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Recently I have noticed Jenkins page rendering being blocked for a long time (sometimes nearly 30s) with stack traces such as</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">"Handling GET /jenkins/ from ... : ... View/index.jelly AdministrativeMonitorsDecorator/footer.jelly" ...
   java.lang.Thread.State: RUNNABLE
	at java.lang.String.&lt;init&gt;(String.java:207)
	at java.lang.String.substring(String.java:1933)
	at net.sf.json.util.JSONTokener.matches(JSONTokener.java:110)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:912)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONArray._fromJSONTokener(JSONArray.java:1131)
	at net.sf.json.JSONArray.fromObject(JSONArray.java:125)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:351)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject._fromString(JSONObject.java:1145)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:162)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:132)
	at hudson.model.UpdateSite.getJSONObject(UpdateSite.java:338)
	at hudson.model.UpdateSite.getData(UpdateSite.java:323)
	at hudson.model.UpdateCenter.getCoreSource(UpdateCenter.java:604)
	at hudson.model.UpdateCenter$CoreUpdateMonitor.getData(UpdateCenter.java:1035)
	at hudson.model.UpdateCenter$CoreUpdateMonitor.isActivated(UpdateCenter.java:1030)
	at jenkins.model.Jenkins.lambda$getActiveAdministrativeMonitors$0(Jenkins.java:2183)
	at jenkins.model.Jenkins$$Lambda$103/1298422401.test(Unknown Source)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
	at jenkins.model.Jenkins.getActiveAdministrativeMonitors(Jenkins.java:2188)
	at jenkins.management.AdministrativeMonitorsDecorator.getActiveAdministrativeMonitors(AdministrativeMonitorsDecorator.java:77)
	at jenkins.management.AdministrativeMonitorsDecorator.getActiveAdministrativeMonitorsCount(AdministrativeMonitorsDecorator.java:72)
	at jenkins.management.AdministrativeMonitorsDecorator.shouldDisplay(AdministrativeMonitorsDecorator.java:131)
</pre>
</div></div>

<p>or</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">  java.lang.Thread.State: RUNNABLE
	at java.util.regex.Pattern.compile(java.base@11.0.5/Pattern.java:1757)
	at java.util.regex.Pattern.&lt;init&gt;(java.base@11.0.5/Pattern.java:1428)
	at java.util.regex.Pattern.compile(java.base@11.0.5/Pattern.java:1068)
	at net.sf.json.regexp.JdkRegexpMatcher.&lt;init&gt;(JdkRegexpMatcher.java:38)
	at net.sf.json.regexp.JdkRegexpMatcher.&lt;init&gt;(JdkRegexpMatcher.java:31)
	at net.sf.json.regexp.RegexpUtils.getMatcher(RegexpUtils.java:39)
	at net.sf.json.util.JSONTokener.matches(JSONTokener.java:111)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:912)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONArray._fromJSONTokener(JSONArray.java:1131)
	at net.sf.json.JSONArray.fromObject(JSONArray.java:125)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:351)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:156)
	at net.sf.json.util.JSONTokener.nextValue(JSONTokener.java:348)
	at net.sf.json.JSONObject._fromJSONTokener(JSONObject.java:955)
	at net.sf.json.JSONObject._fromString(JSONObject.java:1145)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:162)
	at net.sf.json.JSONObject.fromObject(JSONObject.java:132)
</pre>
</div></div>

<p>While loading the JSON text from disk takes just milliseconds, the parsing is extremely slow (hundreds of times slower than <tt>jq</tt>, for example). This seems to be due to poor code in <tt>json-lib</tt>, especially <a href="https://github.com/jenkinsci/json-lib/blame/master/src/main/jdk15/net/sf/json/JSONObject.java#L829" class="external-link" target="_blank" rel="nofollow noopener">this</a> which repeatedly compiles patterns where a simple <tt>.startsWith("null")</tt> would apparently have sufficed!</p>

<p>The poor performance of <tt>JSONObject.fromObject</tt> might not have been so noticeable except that some administrative monitors blocked on the parsing, contrary to the Javadoc in <tt>AdministrativeMonitor.isActivated</tt>.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-60625?focusedId=383003&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383003">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Note that I am still inclined to look for ways to patch or bypass this library to minimize even background CPU, but that would be tracked as a separate issue and presumably not be an LTS candidate.</p>
</pre>
</details>

</details>
<p>Note that I am still inclined to look for ways to patch or bypass this library to minimize even background CPU, but that would be tracked as a separate issue and presumably not be an LTS candidate.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> olivergondza's <a href="https://issues.jenkins.io/browse/JENKINS-60625?focusedId=383011&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383011">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I would like to give this change some more soaking time so postponing to 2.204.3</p>

<p> </p>
</pre>
</details>

</details>
<p>I would like to give this change some more soaking time so postponing to 2.204.3</p>

<p> </p>
